source/common/upstream/health_checker_impl.cc

```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
   request_.asArray().swap(values);
 }
 
+GrpcHealthCheckerImpl::GrpcHealthCheckerImpl(const Cluster& cluster,
+                                             const envoy::api::v2::HealthCheck& config,
+                                             Event::Dispatcher& dispatcher,
+                                             Runtime::Loader& runtime,
+                                             Runtime::RandomGenerator& random)
+    : HealthCheckerImplBase(cluster, config, dispatcher, runtime, random),
+      service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "grpc.health.v1.Health.Check")) {
+  if (!config.grpc_health_check().service_name().empty()) {
+    service_name_.value(config.grpc_health_check().service_name());
+  }
+}
+
+GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::GrpcActiveHealthCheckSession(
+    GrpcHealthCheckerImpl& parent, HostSharedPtr host)
+    : ActiveHealthCheckSession(parent, host), parent_(parent) {}
+
+GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::~GrpcActiveHealthCheckSession() {
+  if (client_) {
+    // If there is an active request it will get reset, so make sure we ignore the reset.
+    expect_reset_ = true;
+    client_->close();
+  }
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::decodeHeaders(
+    Http::HeaderMapPtr&& headers, bool end_stream) {
+  const auto http_response_status = Http::Utility::getResponseStatus(*headers);
+  if (http_response_status != enumToInt(Http::Code::OK)) {
+    // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md requires that
+    // grpc-status be used if available.
+    if (end_stream) {
+      auto grpc_status = Grpc::Common::getGrpcStatus(*headers);
+      if (grpc_status.valid()) {
+        onRpcComplete(grpc_status.value(), Grpc::Common::getGrpcMessage(*headers), true);
```

Comment:
Isn't gRPC always 200? If it's non-200, it seems like we shouldn't be expecting `grpc-status` at all.

Reply:
this code was taken from `grpc_async_client_impl.h`. I haven't managed to receive such a thing from grpc-go server, but maybe it can exist

Reply:
This is copied from [Grpc::AsyncClentImpl](https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h#L111). I wasn't able to get to this codepath from grpc-go server, but maybe it can happen with more intermediaries en route.

Reply:
Ah, fair enough. I actually wrote that code, and suspect that it's gratuitous, but let's leave it as is for consistency.